### PR TITLE
Output buffer middleware

### PR DESCRIPTION
**Pull request by codeguy**
Work in progress. Opening PR for feedback.

### CODE REVIEW

**Comment by codeguy** [comment]
Changes so far include:

1. Require all route callbacks to return PSR7 response.
2. Remove all output buffering concerns from `App` and `Route` classes.

**Review by akrabat — CHANGES_REQUESTED** [review]
* ~Need tests for the new `OutputBuffering` middleware~
* Can you add a CHANGELOG entry please?

Update - I can't read!

File: tests/RouteTest.php [hunk 1]
```diff
@@ -296,36 +276,11 @@ public function testInvokeWhenEchoingOutput()
     public function testInvokeWhenReturningAString()
```
Comment (akrabat) [inline tests/RouteTest.php:276]: I think this test should return a string and expect an Exception to happen?

**Review by akrabat — CHANGES_REQUESTED** [review]

File: Slim/App.php [hunk 2]
```diff
@@ -543,11 +542,6 @@ public function map(array $methods, $pattern, $callable)
         // Create route
         $route = $this->getRouter()->map($methods, $pattern, $callable);
 
-        // TODO: Do we keep output buffering?
-        if (is_callable([$route, 'setOutputBuffering'])) {
-            $route->setOutputBuffering($this->getSetting('outputBuffering', 'append'));
-        }
```
Comment (geggleto) [inline Slim/App.php:549]: I honestly think that we can remove it.

Dump and Die will still work, which is I feel what a majority of people do anyway.

File: tests/RouteTest.php (hunk 1 above)
Comment (geggleto) [reply tests/RouteTest.php:276]: @akrabat I think the name of the test should change. This test should be renamed.

`testEchoingOutputIsEmpty`

`testReturningStringThrowsException` ?

**Review by geggleto — CHANGES_REQUESTED** [review]

File: Slim/Middleware/OutputBuffering.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+     * @param  callable                $next Middleware callable
+     * @return \ResponseInterface            PSR7 response
+     */
+    public function __invoke(ServerRequestInterface $req, ResponseInterface $res, callable $next)
+    {
+        try {
+            ob_start();
+            $newResponse = $next($req, $res);
+            $output = ob_get_clean();
+        } catch (\Throwable $e) {
```
Comment (danopz) [inline Slim/Middleware/OutputBuffering.php:46]: PHP 5 vs PHP 7 #2175

**Pull request merged**